src/signature.c

```diff
@@ -389,17 +389,26 @@ GBytes *cms_sign(GBytes *content, const gchar *certfile, const gchar *keyfile, g
 	}
 
 	/* keyring was given, perform verification to obtain trust chain */
-	if (r_context()->config->keyring_path) {
+	if (r_context()->verify_keyringpath) {
+		gchar *orig_keyring_path;
 		g_autoptr(CMS_ContentInfo) vcms = NULL;
 		g_autoptr(X509_STORE) store = NULL;
 		STACK_OF(X509) *verified_chain = NULL;
 
+		/* temporary set config->keyring_path to verify_keyring_pathpath */
 		g_message("Keyring given, doing signature verification");
+		orig_keyring_path = r_context()->config->keyring_path;
+		r_context()->config->keyring_path = r_context()->verify_keyringpath;
```

Comment:
The context and config is intended to be read-only for the lower layers.

Reply:
I've changed this by adding new arguments to cms_sign() that allow explicitly passing non-default capath/cadir.